nasl/nasl_packet_forgery.c

```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
+
+          // Need reallocated memory because options requires it.
+          if (total_opt_len > opt_size_allocated)
+            {
+              opt_size_allocated = ((total_opt_len / 4) + 1) * 4;
+              opts = g_realloc (opts, sizeof (char) * opt_size_allocated);
+              ptr_opts_pos = opts + current_opt_len;
+            }
+
+          memcpy (ptr_opts_pos, (u_char *) opt_tstamp,
+                  sizeof (struct tcp_opt_tstamp));
+          ptr_opts_pos = ptr_opts_pos + sizeof (struct tcp_opt_tstamp);
+          g_free (opt_tstamp);
+          break;
+        case TCPOPT_NOP:
+        case TCPOPT_EOL:
+        case TCPOPT_SACK: /* Experimental, not supported */
+        default:
+          nasl_perror (lexic, "%s: TCP option %d not supported\n", __func__,
+                       tcp_opt);
+          break;
+        }
+    }
+
+  // Add NOP padding and End Of Option list kinds.
+  current_opt_len = total_opt_len;
+  eol = TCPOPT_EOL;
+  nop = TCPOPT_NOP;
+  if (total_opt_len % 4 == 0)
+    {
+      opt_size_allocated = opt_size_allocated + 4;
+      opts = g_realloc (opts, sizeof (char) * opt_size_allocated);
+      ptr_opts_pos = opts + total_opt_len;
+    }
+  if (current_opt_len < opt_size_allocated - 1)
+    {
+      // Add NOPs
+      for (i = current_opt_len; i < opt_size_allocated - 1; i++)
+        {
+          memcpy (ptr_opts_pos, &nop, 1);
+          total_opt_len++;
+          ptr_opts_pos++;
+        }
+    }
+  // Add EOL
+  memcpy (ptr_opts_pos, &eol, 1);
+
+  if (ip->ip_hl * 4 > pktsz)
+    tcp =
+      (struct tcphdr *) (pkt
+                         + 20); /* ip->ip_hl is bogus, we work around that */
+  else
+    tcp = (struct tcphdr *) (pkt + ip->ip_hl * 4);
+
+  if (pktsz < UNFIX (ip->ip_len))
+    return NULL;
+
+  if (data_len == 0)
+    {
+      data_len = UNFIX (ip->ip_len) - (ip->ip_hl * 4) - (tcp->th_off * 4);
+      data = (char *) ((char *) tcp + tcp->th_off * 4);
+    }
+
+  // Alloc enough memory to hold the options
+  npkt =
+    g_malloc0 (ip->ip_hl * 4 + tcp->th_off * 4 + opt_size_allocated + data_len);
+  memcpy (npkt, pkt, UNFIX (ip->ip_len));
+  ip = (struct ip *) (npkt);
+  tcp = (struct tcphdr *) (npkt + ip->ip_hl * 4);
+
+  // copy options
+  memcpy ((char *) tcp + tcp->th_off * 4, opts, opt_size_allocated);
+
+  tcp->th_off = tcp->th_off + (opt_size_allocated / 4);
+  memcpy ((char *) tcp + tcp->th_off * 4, data, data_len);
+
+  // Update ip_len and calculate ip checksum
+  ip->ip_len = FIX (ip->ip_hl * 4 + tcp->th_off * 4 + data_len);
+  ip->ip_sum = 0;
+  ip->ip_sum = np_in_cksum ((u_short *) npkt, ip->ip_hl * 4);
+
+  // Calculate tcp header with options checksum
+  struct pseudohdr pseudoheader;
+  char *tcpsumdata =
+    g_malloc0 (sizeof (struct pseudohdr) + opt_size_allocated + data_len + 1);
+  struct in_addr source, dest;
+
+  source.s_addr = ip->ip_src.s_addr;
+  dest.s_addr = ip->ip_dst.s_addr;
+
+  bzero (&pseudoheader, sizeof (pseudoheader));
```

Comment:
bzero and bcopy are deprecated. memset and memcpy should be used instead. Same for all following occurrences.